torch/_lobpcg.py

```diff
@@ -13,6 +13,336 @@
+    # Note that `chr_poly_D_at_A_to_U_ortho` is positive-definite if
+    # 1. `largest` == False, or
+    # 2. `largest` == True and `k` is even
+    # under the assumption that `A` has distinct eigenvalues.
+    #
+    # check if `chr_poly_D_at_A_to_U_ortho` is positive-definite or negative-definite
+    chr_poly_D_at_A_to_U_ortho_sign = -1 if (largest and (k % 2 == 1)) else +1
+    chr_poly_D_at_A_to_U_ortho_L = torch.cholesky(
+        chr_poly_D_at_A_to_U_ortho_sign * chr_poly_D_at_A_to_U_ortho
+    )
+
+    # compute the gradient part in span(U)
+    res = _symeig_backward_complete_eigenspace(
+        D_grad, U_grad, A, D, U
+    )
+
+    # incorporate the Sylvester equation solution into the full gradient
+    # it resides in span(U_ortho)
+    res -= U_ortho.matmul(
+        chr_poly_D_at_A_to_U_ortho_sign * torch.cholesky_solve(
+            U_ortho_t.matmul(series_acc),
+            chr_poly_D_at_A_to_U_ortho_L
+        )
+    ).matmul(Ut)
+
+    return res
+
+def _symeig_backward(D_grad, U_grad, A, D, U, largest):
+    # if `U` is square, then the columns of `U` is a complete eigenspace
+    if U.size(-1) == U.size(-2):
+        return _symeig_backward_complete_eigenspace(
+            D_grad, U_grad, A, D, U
+        )
+    else:
+        return _symeig_backward_partial_eigenspace(
+            D_grad, U_grad, A, D, U, largest
+        )
+
+class LOBPCGAutogradFunction(torch.autograd.Function):
+
+    @staticmethod
+    def forward(ctx,
+                A,              # type: Tensor
+                k,              # type: Optional[int]
+                B,              # type: Optional[Tensor]
+                X,              # type: Optional[Tensor]
+                n,              # type: Optional[int]
+                iK,             # type: Optional[Tensor]
+                niter,          # type: Optional[int]
+                tol,            # type: Optional[float]
+                largest,        # type: Optional[bool]
+                method,         # type: Optional[str]
+                tracker,        # type: Optional[None]
+                ortho_iparams,  # type: Optional[Dict[str, int]]
+                ortho_fparams,  # type: Optional[Dict[str, float]]
+                ortho_bparams,  # type: Optional[Dict[str, bool]]
+                ):
+        # type: (...) -> Tuple[Tensor, Tensor]
+
+        # makes sure that input is contiguous for efficiency.
+        # Note: autograd does not support dense gradients for sparse input yet.
+        A = A.contiguous() if (not A.is_sparse) else A
+        if B is not None:
+            B = B.contiguous() if (not B.is_sparse) else B
+
+        D, U = _lobpcg(
+            A, k, B, X,
+            n, iK, niter, tol, largest, method, tracker,
+            ortho_iparams, ortho_fparams, ortho_bparams
+        )
+        # D, U = torch.symeig(A, eigenvectors=True, upper=True)
+        # D = D[..., -k:]
+        # U = U[..., :, -k:]
+
+        D = D.contiguous()
```

Comment:
Is this actually needed?

Reply:
Not needed, has kept since the `symeig` in forward. Will remove.